# 4.8 TASK 4.8 Grouping data and aggregating variables

# Table of Contents

#### 01. Import libraries

#### 02. Delete column

#### 03. Delete records impacted by outliers

#### 04. Export dataframe as pickle file

# 01. Import libraries

In [4]:
# Import libraries

import pandas as pd
import numpy as np
import os

# Step 1

In [7]:
# Import dataframe

path = r'C:\Users\dirk8\CareerFoundry Projects\03-2025 Instacart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'ords_prods_merge_3.pkl'))

In [9]:
# Check the output
ords_prods_merge[53:62]

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc,busiest_day,busiest_days,busiest_period_of_day,max_order,loyalty_flag
53,2550362,1,10,4,8,30.0,38928,4,0,0% Greek Strained Yogurt,120,16,12.6,Mid-range product,Least busy,Slowest days,Average orders,10,New customer
54,2550362,1,10,4,8,30.0,25133,5,1,Organic String Cheese,21,16,8.6,Mid-range product,Least busy,Slowest days,Average orders,10,New customer
55,2550362,1,10,4,8,30.0,10258,6,1,Pistachios,117,19,3.0,Low-range product,Least busy,Slowest days,Average orders,10,New customer
56,2550362,1,10,4,8,30.0,35951,7,0,Organic Unsweetened Almond Milk,91,16,1.3,Low-range product,Least busy,Slowest days,Average orders,10,New customer
57,2550362,1,10,4,8,30.0,13032,8,1,Cinnamon Toast Crunch,121,14,4.0,Low-range product,Least busy,Slowest days,Average orders,10,New customer
58,2550362,1,10,4,8,30.0,12427,9,1,Original Beef Jerky,23,19,4.4,Low-range product,Least busy,Slowest days,Average orders,10,New customer
59,2168274,2,1,2,11,NaN,32792,1,0,Chipotle Beef & Pork Realstick,23,19,5.2,Mid-range product,Regularly busy,Regularly busy,Most orders,14,Regular customer
60,2168274,2,1,2,11,NaN,47766,2,0,Organic Avocado,24,4,6.3,Mid-range product,Regularly busy,Regularly busy,Most orders,14,Regular customer
61,2168274,2,1,2,11,NaN,20574,3,0,Roasted Turkey,96,20,2.3,Low-range product,Regularly busy,Regularly busy,Most orders,14,Regular customer


In [11]:
# Check dimensions
ords_prods_merge.shape

(32404859, 19)

# 02. Delete column

In [13]:
# Tutor feedback confirmed that the 'busiest_days' column is not needed, so delete it
# Dataframe before change is stored locally (as the imported ords_prods_merge_3.pkl) as backup

ords_prods_merge.drop('busiest_days', axis=1, inplace=True)

In [15]:
# Check the output
ords_prods_merge[53:62]

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc,busiest_day,busiest_period_of_day,max_order,loyalty_flag
53,2550362,1,10,4,8,30.0,38928,4,0,0% Greek Strained Yogurt,120,16,12.6,Mid-range product,Least busy,Average orders,10,New customer
54,2550362,1,10,4,8,30.0,25133,5,1,Organic String Cheese,21,16,8.6,Mid-range product,Least busy,Average orders,10,New customer
55,2550362,1,10,4,8,30.0,10258,6,1,Pistachios,117,19,3.0,Low-range product,Least busy,Average orders,10,New customer
56,2550362,1,10,4,8,30.0,35951,7,0,Organic Unsweetened Almond Milk,91,16,1.3,Low-range product,Least busy,Average orders,10,New customer
57,2550362,1,10,4,8,30.0,13032,8,1,Cinnamon Toast Crunch,121,14,4.0,Low-range product,Least busy,Average orders,10,New customer
58,2550362,1,10,4,8,30.0,12427,9,1,Original Beef Jerky,23,19,4.4,Low-range product,Least busy,Average orders,10,New customer
59,2168274,2,1,2,11,NaN,32792,1,0,Chipotle Beef & Pork Realstick,23,19,5.2,Mid-range product,Regularly busy,Most orders,14,Regular customer
60,2168274,2,1,2,11,NaN,47766,2,0,Organic Avocado,24,4,6.3,Mid-range product,Regularly busy,Most orders,14,Regular customer
61,2168274,2,1,2,11,NaN,20574,3,0,Roasted Turkey,96,20,2.3,Low-range product,Regularly busy,Most orders,14,Regular customer


In [17]:
# Check the dimensions
ords_prods_merge.shape

(32404859, 18)

# 03. Delete records impacted by outliers

In [19]:
# Tutor gave me a green light to delete (or impute values) for large outliers in the 'prices' column that may be skewing mean and standard deviation
# These outliers were discovered recently rather than at the original consistency check.  Better late than never.
# To recap, showing descriptive stats on the 'prices' column

ords_prods_merge['prices'].describe()

count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

__The max of 99999 and standard deviation of 496 (rounded), against a mean of 12 (rounded), confirm outlier-caused heavy skew__

In [25]:
# Investigating products priced above 10000

expensive_items = ords_prods_merge[ords_prods_merge['prices'] > 10000]

In [29]:
# Check the dimensions

expensive_items.shape

(5127, 18)

In [43]:
# Investigate a few

expensive_items.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc,busiest_day,busiest_period_of_day,max_order,loyalty_flag
1576,912404,17,12,2,14,5.0,21553,5,0,Lowfat 2% Milkfat Cottage Cheese,108,16,14900.0,High-range product,Regularly busy,Most orders,40,Regular customer
1638,603376,17,22,6,16,4.0,21553,3,1,Lowfat 2% Milkfat Cottage Cheese,108,16,14900.0,High-range product,Regularly busy,Most orders,40,Regular customer
16522,3264360,135,2,2,21,13.0,21553,6,0,Lowfat 2% Milkfat Cottage Cheese,108,16,14900.0,High-range product,Regularly busy,Fewest orders,4,New customer
16528,892534,135,3,0,8,12.0,21553,3,1,Lowfat 2% Milkfat Cottage Cheese,108,16,14900.0,High-range product,Busiest day,Average orders,4,New customer
53672,229704,342,8,1,19,30.0,21553,9,0,Lowfat 2% Milkfat Cottage Cheese,108,16,14900.0,High-range product,Regularly busy,Average orders,16,Regular customer


In [45]:
# Investigate a few more

expensive_items.tail()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc,busiest_day,busiest_period_of_day,max_order,loyalty_flag
32320511,3172853,205650,18,1,9,7.0,21553,17,1,Lowfat 2% Milkfat Cottage Cheese,108,16,14900.0,High-range product,Regularly busy,Most orders,25,Regular customer
32347717,2504315,205818,3,5,15,3.0,21553,13,0,Lowfat 2% Milkfat Cottage Cheese,108,16,14900.0,High-range product,Regularly busy,Most orders,25,Regular customer
32347727,1108388,205818,5,4,5,1.0,21553,5,1,Lowfat 2% Milkfat Cottage Cheese,108,16,14900.0,High-range product,Least busy,Fewest orders,25,Regular customer
32380527,1916142,206049,1,2,17,NaN,21553,2,0,Lowfat 2% Milkfat Cottage Cheese,108,16,14900.0,High-range product,Regularly busy,Most orders,5,New customer
32380551,379732,206049,4,1,14,5.0,21553,4,1,Lowfat 2% Milkfat Cottage Cheese,108,16,14900.0,High-range product,Regularly busy,Most orders,5,New customer


In [47]:
# Show stats

expensive_items.describe()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,aisle_id,department_id,prices,max_order
count,5.127000e+03,5127.000000,5127.000000,5127.000000,5127.000000,4776.000000,5127.000000,5127.000000,5127.000000,5127.000000,5127.0,5127.000000,5127.000000
mean,1.708876e+06,106117.185684,15.302126,2.750536,13.473376,11.899079,23201.815682,7.986347,0.613614,104.732592,16.0,26485.547494,29.791691
std,9.838956e+05,58594.982659,16.691849,2.010892,4.137326,9.068019,4153.742747,6.868889,0.486968,8.231346,0.0,29186.636448,24.929207
min,4.300000e+01,17.000000,1.000000,0.000000,0.000000,0.000000,21553.000000,1.000000,0.000000,84.000000,16.0,14900.000000,3.000000
25%,8.642320e+05,58136.000000,4.000000,1.000000,10.000000,5.000000,21553.000000,3.000000,0.000000,108.000000,16.0,14900.000000,12.000000
50%,1.693799e+06,106021.000000,9.000000,3.000000,13.000000,8.000000,21553.000000,6.000000,1.000000,108.000000,16.0,14900.000000,21.000000
75%,2.567244e+06,156389.000000,20.000000,5.000000,16.000000,16.000000,21553.000000,11.000000,1.000000,108.000000,16.0,14900.000000,41.000000
max,3.421047e+06,206049.000000,99.000000,6.000000,23.000000,30.000000,33664.000000,72.000000,1.000000,108.000000,16.0,99999.000000,99.000000


In [49]:
# Show frequency counts

expensive_items['prices'].value_counts()

prices
14900.0    4429
99999.0     698
Name: count, dtype: int64

In [51]:
# Show product_name frequency counts to investigate further

expensive_items['product_name'].value_counts()

product_name
Lowfat 2% Milkfat Cottage Cheese    4429
2 % Reduced Fat  Milk                698
Name: count, dtype: int64

__Reviewing the 2 tables above, 14900 and 99999 are clearly inaccurate prices for cottage cheese and milk, respectively.__

In [55]:
# Check the dimensions

expensive_items.shape

(5127, 18)

In [57]:
# Export the created expensive_items subset dataframe for backup, prior to deleting these records from the ords_prods_merge dataframe

expensive_items.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'expensive_items.pkl'))

# Correct the 'prices' column data significant inaccuracies via deletion of impacted records

In [61]:
# Check dimensions of ords_prods_merge
ords_prods_merge.shape

(32404859, 18)

In [68]:
# Delete rows where 'prices' column value equals 14900 or 99999 from the dataframe, to correct significant price inaccuracies that skew analysis

ords_prods_merge_1 = ords_prods_merge[ords_prods_merge['prices'].isin([14900, 99999]) == False]

In [70]:
# Check the dimensions

ords_prods_merge_1.shape

(32399732, 18)

In [72]:
# Check the output

ords_prods_merge_1.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc,busiest_day,busiest_period_of_day,max_order,loyalty_flag
0,2539329,1,1,2,8,NaN,196,1,0,Soda,77,7,9.0,Mid-range product,Regularly busy,Average orders,10,New customer
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5,Mid-range product,Regularly busy,Average orders,10,New customer
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4,Low-range product,Regularly busy,Average orders,10,New customer
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7,Low-range product,Regularly busy,Average orders,10,New customer
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0,Low-range product,Regularly busy,Average orders,10,New customer


In [80]:
# Check whether 'prices' condition caused relevant records to be deleted

ords_prods_merge_1[ords_prods_merge_1['prices'] > 10000]

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc,busiest_day,busiest_period_of_day,max_order,loyalty_flag


In [74]:
# Check stats of original dataframe

ords_prods_merge.describe()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,aisle_id,department_id,prices,max_order
count,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.032876e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07,3.240486e+07
mean,1.710745e+06,1.029372e+05,1.714230e+01,2.738867e+00,1.342515e+01,1.110408e+01,2.559866e+04,8.352547e+00,5.895873e-01,7.119612e+01,9.919792e+00,1.198023e+01,3.305217e+01
std,9.872988e+05,5.946610e+04,1.753532e+01,2.090077e+00,4.246380e+00,8.779064e+00,1.408400e+04,7.127071e+00,4.919087e-01,3.821139e+01,6.281485e+00,4.956554e+02,2.515525e+01
min,2.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00
25%,8.559470e+05,5.142200e+04,5.000000e+00,1.000000e+00,1.000000e+01,5.000000e+00,1.354400e+04,3.000000e+00,0.000000e+00,3.100000e+01,4.000000e+00,4.200000e+00,1.300000e+01
50%,1.711049e+06,1.026160e+05,1.100000e+01,3.000000e+00,1.300000e+01,8.000000e+00,2.530200e+04,6.000000e+00,1.000000e+00,8.300000e+01,9.000000e+00,7.400000e+00,2.600000e+01
75%,2.565499e+06,1.543890e+05,2.400000e+01,5.000000e+00,1.600000e+01,1.500000e+01,3.794700e+04,1.100000e+01,1.000000e+00,1.070000e+02,1.600000e+01,1.130000e+01,4.700000e+01
max,3.421083e+06,2.062090e+05,9.900000e+01,6.000000e+00,2.300000e+01,3.000000e+01,4.968800e+04,1.450000e+02,1.000000e+00,1.340000e+02,2.100000e+01,9.999900e+04,9.900000e+01


In [76]:
# Check the dimensions

ords_prods_merge.shape

(32404859, 18)

In [82]:
# Export ords_prods_merge dataframe as backup before deletions were made

ords_prods_merge.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'ords_prods_merge_4.pkl'))

In [5]:
# Import libraries and dataframe

import pandas as pd
import numpy as np
import os

path = r'C:\Users\dirk8\CareerFoundry Projects\03-2025 Instacart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'ords_prods_merge_4.pkl'))

In [7]:
# Check the dimensions

ords_prods_merge.shape

(32404859, 18)

In [9]:
# Check the output

ords_prods_merge.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc,busiest_day,busiest_period_of_day,max_order,loyalty_flag
0,2539329,1,1,2,8,NaN,196,1,0,Soda,77,7,9.0,Mid-range product,Regularly busy,Average orders,10,New customer
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5,Mid-range product,Regularly busy,Average orders,10,New customer
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4,Low-range product,Regularly busy,Average orders,10,New customer
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7,Low-range product,Regularly busy,Average orders,10,New customer
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0,Low-range product,Regularly busy,Average orders,10,New customer


In [11]:
# Delete rows where 'prices' column value < 14900

ords_prods_merge = ords_prods_merge[ords_prods_merge['prices'] < 14900]

In [13]:
# Check the dimensions

ords_prods_merge.shape

(32399732, 18)

In [15]:
# Check the stats

ords_prods_merge.describe()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,aisle_id,department_id,prices,max_order
count,3.239973e+07,3.239973e+07,3.239973e+07,3.239973e+07,3.239973e+07,3.032399e+07,3.239973e+07,3.239973e+07,3.239973e+07,3.239973e+07,3.239973e+07,3.239973e+07,3.239973e+07
mean,1.710746e+06,1.029367e+05,1.714260e+01,2.738865e+00,1.342514e+01,1.110395e+01,2.559904e+04,8.352604e+00,5.895835e-01,7.119081e+01,9.918830e+00,7.790994e+00,3.305268e+01
std,9.872994e+05,5.946623e+04,1.753544e+01,2.090089e+00,4.246397e+00,8.779012e+00,1.408498e+04,7.127110e+00,4.919093e-01,3.821194e+01,6.281516e+00,4.241809e+00,2.515525e+01
min,2.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00
25%,8.559450e+05,5.142000e+04,5.000000e+00,1.000000e+00,1.000000e+01,5.000000e+00,1.354100e+04,3.000000e+00,0.000000e+00,3.100000e+01,4.000000e+00,4.200000e+00,1.300000e+01
50%,1.711052e+06,1.026140e+05,1.100000e+01,3.000000e+00,1.300000e+01,8.000000e+00,2.530500e+04,6.000000e+00,1.000000e+00,8.300000e+01,9.000000e+00,7.400000e+00,2.600000e+01
75%,2.565499e+06,1.543880e+05,2.400000e+01,5.000000e+00,1.600000e+01,1.500000e+01,3.794700e+04,1.100000e+01,1.000000e+00,1.070000e+02,1.600000e+01,1.130000e+01,4.700000e+01
max,3.421083e+06,2.062090e+05,9.900000e+01,6.000000e+00,2.300000e+01,3.000000e+01,4.968800e+04,1.450000e+02,1.000000e+00,1.340000e+02,2.100000e+01,2.500000e+01,9.900000e+01


__The 'prices' column stats now look good, with a maximum of 25, minimum of 1, mean of 7.79 and standard deviation of 4.24__

In [17]:
# Check the output

ords_prods_merge.head()

,order_id,user_id,order_number,orders_day_of_week,order_hour_of_day,days_since_prior_order,product_id,add_to_cart_order,reordered,product_name,aisle_id,department_id,prices,price_range_loc,busiest_day,busiest_period_of_day,max_order,loyalty_flag
0,2539329,1,1,2,8,NaN,196,1,0,Soda,77,7,9.0,Mid-range product,Regularly busy,Average orders,10,New customer
1,2539329,1,1,2,8,NaN,14084,2,0,Organic Unsweetened Vanilla Almond Milk,91,16,12.5,Mid-range product,Regularly busy,Average orders,10,New customer
2,2539329,1,1,2,8,NaN,12427,3,0,Original Beef Jerky,23,19,4.4,Low-range product,Regularly busy,Average orders,10,New customer
3,2539329,1,1,2,8,NaN,26088,4,0,Aged White Cheddar Popcorn,23,19,4.7,Low-range product,Regularly busy,Average orders,10,New customer
4,2539329,1,1,2,8,NaN,26405,5,0,XL Pick-A-Size Paper Towel Rolls,54,17,1.0,Low-range product,Regularly busy,Average orders,10,New customer


# 04. Export dataframe as pickle file

In [19]:
# Export the dataframe

ords_prods_merge.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'ords_prods_merge_5.pkl'))